File: lora/training/ragnhild/ai_toolkit/check_hf_flux_access.py
#!/usr/bin/env python3
"""Preflight check for Ragnhild FLUX LoRA train/generate jobs."""

from __future__ import annotations

import os
import sys


FLUX_REPO = os.environ.get("RAGNHILD_FLUX_MODEL", "black-forest-labs/FLUX.1-dev")
FLUX_LICENSE_URL = f"https://huggingface.co/{FLUX_REPO}"


def token() -> str | None:
    for key in ("HF_TOKEN", "HUGGINGFACE_HUB_TOKEN"):
        value = os.environ.get(key, "").strip()
        if value:
            return value

    for path in (
        os.path.expanduser("~/.cache/huggingface/token"),
        os.path.expanduser("~/.huggingface/token"),
    ):
        if os.path.isfile(path):
            with open(path, encoding="utf-8") as handle:
                value = handle.read().strip()
            if value:
                return value
    return None


def local_flux_path() -> str | None:
    override = os.environ.get("RAGNHILD_FLUX_MODEL_PATH", "").strip()
    if override and os.path.isdir(override):
        return override
    return None


def main() -> int:
    local_path = local_flux_path()
    if local_path:
        print(f"OK: using local FLUX model at {local_path}")
        return 0

    auth = token()
    if not auth:
        print("BLOCKED: Hugging Face auth not found.", file=sys.stderr)
        print("Set HF_TOKEN or run: hf auth login", file=sys.stderr)
        return 1

    try:
        from huggingface_hub import HfApi, hf_hub_download
    except ImportError:
        print("BLOCKED: huggingface_hub is not installed in the active Python env.", file=sys.stderr)
        return 1

    api = HfApi(token=auth)
    try:
        user = api.whoami()["name"]
        print(f"HF user: {user}")
    except Exception as exc:  # noqa: BLE001
        print(f"BLOCKED: Hugging Face token is invalid ({exc}).", file=sys.stderr)
        return 1

    try:
        path = hf_hub_download(FLUX_REPO, "model_index.json", token=auth)
        print(f"OK: authorized for {FLUX_REPO}")
        print(f"model_index: {path}")
        return 0
    except Exception as exc:  # noqa: BLE001
        message = str(exc)
        print(f"BLOCKED: cannot download {FLUX_REPO}", file=sys.stderr)
        if "GatedRepoError" in type(exc).__name__ or "403" in message or "authorized list" in message:
            print("Reason: account is logged in but not approved for this gated model.", file=sys.stderr)
            print(f"Fix: open {FLUX_LICENSE_URL} while logged in as {user}, click Agree, then rerun.", file=sys.stderr)
        else:
            print(f"Reason: {exc}", file=sys.stderr)
        print("Alternative: set RAGNHILD_FLUX_MODEL_PATH to a local FLUX.1-dev checkout.", file=sys.stderr)
        return 2


if __name__ == "__main__":
    raise SystemExit(main())

File: STUDIO/lora/_toolkit/check_hf_flux_access.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require "net/http"
require "json"
require "uri"

# Nothing to check when FLUX is not the base.
#
# This gate exists because FLUX.1-dev is licence-gated: a token that has not
# accepted the terms downloads a 403 and the failure appears hundreds of lines
# later inside diffusers. Real problem, and entirely FLUX's.
#
# SDXL is not gated. It needs no token at all. But every lane in run_generate.sh
# calls this gate unconditionally, so an SDXL render on a machine with no HF
# credentials refused to start — blocked by the licence terms of a model it was
# never going to load.
#
# Exits 0 with a note rather than skipping silently, because a gate that stops
# reporting is indistinguishable from a gate that stopped mattering.
if ENV["LORA_BASE"].to_s.strip.downcase == "sdxl"
  puts "ok: base is SDXL, which is not licence-gated — no Hugging Face token needed"
  exit 0
end

FLUX_REPO = ENV.fetch("LORA_FLUX_MODEL", "black-forest-labs/FLUX.1-dev")
FLUX_URL = "https://huggingface.co/#{FLUX_REPO}"
TOKEN_PATHS = [
  File.expand_path("~/.cache/huggingface/token"),
  File.expand_path("~/.huggingface/token"),
].freeze

def read_token
  %w[HF_TOKEN HUGGINGFACE_HUB_TOKEN].each do |key|
    value = ENV.fetch(key, "").strip
    return value unless value.empty?
  end

  TOKEN_PATHS.each do |path|
    next unless File.file?(path)

    value = File.read(path, encoding: "UTF-8").strip
    return value unless value.empty?
  end

  nil
end

def local_flux_path
  path = ENV.fetch("LORA_FLUX_MODEL_PATH", "").strip
  path.empty? || !File.directory?(path) ? nil : path
end

def hf_get(url, token)
  uri = URI(url)
  Net::HTTP.start(uri.host, uri.port, use_ssl: uri.scheme == "https", open_timeout: 20, read_timeout: 60) do |http|
    request = Net::HTTP::Get.new(uri)
    request["Authorization"] = "Bearer #{token}" unless token.to_s.empty?
    http.request(request)
  end
end

def gated?(response)
  body = response.body.to_s
  response.code == "403" || body.include?("authorized list") || body.include?("gated")
end

def main
  local = local_flux_path
  if local
    puts "ok: local FLUX at #{local}"
    return 0
  end

  token = read_token
  unless token
    warn "warn: Hugging Face auth missing"
    warn "fix: set HF_TOKEN or run hf auth login"
    return 1
  end

  whoami = hf_get("https://huggingface.co/api/whoami-v2", token)
  unless whoami.code.to_i.between?(200, 299)
    warn "warn: invalid Hugging Face token (#{whoami.code})"
    return 1
  end

  user = JSON.parse(whoami.body).fetch("name", "unknown")
  puts "ok: HF user #{user}"

  model_index_url = "https://huggingface.co/#{FLUX_REPO}/resolve/main/model_index.json"
  index = hf_get(model_index_url, token)
  if index.code.to_i.between?(200, 299)
    puts "ok: authorized for #{FLUX_REPO}"
    puts "ok: model_index #{model_index_url}"
    return 0
  end

  warn "warn: cannot access #{FLUX_REPO}"
  if gated?(index)
    warn "fix: open #{FLUX_URL} as #{user}, accept license, rerun"
  else
    warn "reason: HTTP #{index.code} #{index.body.to_s.strip[0, 200]}"
  end
  warn "fix: or set LORA_FLUX_MODEL_PATH to local FLUX.1-dev"
  2
end

exit(main) if $PROGRAM_NAME == __FILE__
